tools/changelog/changelog/markdown_test.go

```diff
@@ -0,0 +1,98 @@
+package changelog
+
+import (
+	"math/rand"
+	"testing"
+	"time"
+
+	"github.com/hexops/autogold"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestMarkdown(t *testing.T) {
+	t.Parallel()
+
+	changes := []Change{{
+		Type: "improvement",
+		Scope: Scope{
+			Primary:   "cli",
+			SubScopes: []string{"engine", "backend"},
+		},
+		Description: "Foo bar baz",
+		GitHubMeta: GitHubMeta{
+			PullRequestNumbers: []int{9001},
+		},
+	}, {
+		Type: "improvement",
+		Scope: Scope{
+			Primary:   "sdk",
+			SubScopes: []string{"go"},
+		},
+		Description: "Make SDK go brrrr",
+		GitHubMeta: GitHubMeta{
+			PullRequestNumbers: []int{10000},
+		},
+	}, {
+		Type: "fix",
+		Scope: Scope{
+			Primary:   "sdkgen",
+			SubScopes: []string{"nodejs"},
+		},
+		Description: "Fix Typescript SDK code generation.",
+		GitHubMeta: GitHubMeta{
+			PullRequestNumbers: []int{20000},
+		},
+	}, {
+		Type: "fix",
+		Scope: Scope{
+			Primary:   "sdkgen",
+			SubScopes: []string{"go"},
+		},
+		Description: "Fix Go SDK code generation",
+		GitHubMeta: GitHubMeta{
+			PullRequestNumbers: []int{20001},
+		},
+	}}
+
+	rand.Seed(time.Now().UnixNano())
+	rand.Shuffle(len(changes), func(i1, i2 int) {
+		changes[i1], changes[i2] = changes[i2], changes[i1]
+	})
+
+	changelog := Changelog{
+		Changes: changes,
+	}
+
+	version := "3.39.0"
+	date := "2022-08-19"
+	config := MustParseConfig(t, []byte(`
+types:
+  improvement: Improvements
+  fix: Bug Fixes
+`))
+
+	// Put improvements and fixes first:
+	buf, err := changelog.Markdown(config, version, date)
```

Comment:
So currently the whole Changelog has one version, how does this play with
A) Currently CHANGELOG_PENDING.md doesn't emit a version line because goreleaser adds it anyway
B) Rendering CHANGELOG.md in full for the release, which needs to write out every version done so far

Reply:
The draft GitHub release won't say unreleased, it'll say `## v1.2.3 (date)` or some-such at the top. That version comes from the .version file that tracks the next intended release version, and the changelog in the draft GH release is rendered from all of the files in /pending.

The release's text then gets inserted into CHANGELOG.md in the post-release PR, and the automation there is extremely simple, it uses head/tail/cat to insert the new release notes below the heading.

Reply:
A) That doesn't match the changelogs today, so may need some goreleaser changes as well
B) I'd like to see about _not_ needing a post release PR at all

Reply:
A) That doesn't match the changelogs today, so may need some goreleaser changes as well 

Yes, that's right. We'll be replacing much of goreleaser's automation because it's a bit backwards for our purposes. We want to be able to produce the final tarballs/zip files and integration test those *before* releasing/publish. By default, goreleaser doesn't create those archives except as *part* of the release/publish step.

B) I don't think that's possible. Something needs to update CHANGELOG.md and something needs to bump versions. Let's chat about this shortly?